Implement automatic JSON API generation

G10h4ck opened the pull request:
qmake file add jsonapi-generator target to compile JSON API generator
qmake files add rs_jsonapi CONFIG option to enable/disable JSON API at compile
  time
RsTypeSerializer pass down same serialization flags when creating new context
  for nested objects serial job
RsGxsChannels expose a few methods through JSON API as example
Derive a few GXS types (RsGxsChannelGroup, RsGxsChannelPost, RsGxsFile,
  RsMsgMetaData) from RsSerializables so they can be used for the JSON API
Create RsGenericSerializer::SERIALIZATION_FLAG_YIELDING so JSON objects that
  miss some fields can be still deserialized, this improve API usability
SerializeContext offer friendly constructor with default paramethers
Add restbed 4.6 library as git submodule as most systems doesn't have it yet
Add a bit of documentation about JSON API into jsonapi-generator/README.adoc
Add JsonApiServer class to expose the JSON API via HTTP protocol

G10h4ck committed:
Implement automatic JSON API generation

qmake file add jsonapi-generator target to compile JSON API generator
qmake files add rs_jsonapi CONFIG option to enable/disable JSON API at compile
  time
RsTypeSerializer pass down same serialization flags when creating new context
  for nested objects serial job
RsGxsChannels expose a few methods through JSON API as example
Derive a few GXS types (RsGxsChannelGroup, RsGxsChannelPost, RsGxsFile,
  RsMsgMetaData) from RsSerializables so they can be used for the JSON API
Create RsGenericSerializer::SERIALIZATION_FLAG_YIELDING so JSON objects that
  miss some fields can be still deserialized, this improve API usability
SerializeContext offer friendly constructor with default paramethers
Add restbed 4.6 library as git submodule as most systems doesn't have it yet
Add a bit of documentation about JSON API into jsonapi-generator/README.adoc
Add JsonApiServer class to expose the JSON API via HTTP protocol

G10h4ck committed:
Remove unused misleading GXS request status

Something must be really strange to be finished and incomplete at same
  time..

G10h4ck committed:
RsTokenService use enum for request status

G10h4ck committed:
Convert to RsTokenService::GxsRequestStatus

Indicate GxsRequest status with an enum instead of uint_* that make the
  code less readable and more prone to errors

G10h4ck committed:
Expose tokenservice methods trough GxsIfaceHelper

G10h4ck committed:
Complete GxsChannels JSON API with blocking methods

RsGxsIfaceHelper::requestStatus expose it to JSON API
Implemented RsGxsIfaceHelper::waitToken to wait for GXS operations
RsItem::serial_process fix doxygen warning as it's a comment not documentation
RsTypeSerializer add JSON conversion methods for double [de]serialization not
	implemented yet
RsTypeSerializer_PRIVATE_{FROM,TO}_JSON_ARRAY fix doxygen warning as it's a
	comment not documentation
make GxsChannels::ExtraFileHash a bit more reasonable
jsonapi-generator fix unused variable warning if there is no input or output
	paramethers
jsonapi-generator fix generation for inerithed jsonapi methods
Convert to RsSerializable some Gxs structs for the JSON API

G10h4ck committed:
Merge branch 'master' into jsonapi

G10h4ck committed:
jsonapi-generator fix support for void methods

Thanks sehraf for reporting
retroshare://forum?name=fucking%20genius&id=8fd22bd8f99754461e7ba1ca8a727995&msgid=503d75bf7ed7fa7568eeae4db5c8d31a7e124c98

G10h4ck committed:
Add login JSON API

Implement a "normal" way to login RsLoginHelper::attemptLogin
Implement a way to get locations list RsLoginHelper::getLocations
Enable JSON API into retroshare-android-service

G10h4ck committed:
Extend JSON API with reasonable way to create location

G10h4ck committed:
Extend JSON API to shutdown retroshare gracefully

sehraf committed:
add simple chat support

G10h4ck committed:
Merge fixes to serialization from master branch

G10h4ck committed:
Expose some RsFiles method trought JSON API

G10h4ck committed:
Experiment of notification via NotifyClient + JSON API

G10h4ck committed:
Merge branch 'master' into jsonapi

csoler committed:
disabled compilation of cmark until it gets fixed

csoler committed:
disabled compilation of cmark until it gets fixed

sehraf committed:
fix jsonapi.h

G10h4ck committed:
Fix deep search returning 0 results

G10h4ck committed:
Merge pull request #1 from sehraf/jsonapi

add chat and fix compilation

G10h4ck committed:
RsGxsNetService::receiveTurtleSearchResults fix deadlock

Avoid notifing observer while the mutex is locked by storing results in
a temporary set

G10h4ck committed:
Implement JSON API generation for async API calls

Move JSON helpers to util/rsjson.* for better usability
Implement JSON ostream manipulator to print compact and pretty JSON
Use lambdas for API wrappers, integrate better and avoid namespace
  pollution
Removed experimental JSON API for notify client wrapper, notifications
  can be implemented automatically with moderns async API calls
Implement and automatically expose to JSON API
  RsGxsChannels::turtleSearchRequest(
        const std::string& matchString,
        const std::function<void (const RsGxsGroupSummary&)>& multiCallback,
        std::time_t maxWait )

G10h4ck committed:
Merge branch 'jsonapi' of github.com:G10h4ck/RetroShare into jsonapi

sehraf commented:
maybe b7f5d4286f166d25dbf812e9ec2826a96ebdf10c should be cherry-picked by @csoler

G10h4ck committed:
Merge branch 'master' into jsonapi

G10h4ck committed:
Remove #pragma once in .cc file

It was probably introduced by copy paste error on license cleanup

G10h4ck committed:
RsLoginHelper::createLocation expose capability of creating hidden locations

G10h4ck commented:
I think we should now considering this for merge

sehraf commented:
Is the restbed issue solved?
An automatic detection of a system installed version would be nice, too

sehraf commented:
I'd like to emphasize that this pr reduces (de)serialization of anything to a few lines :astonished:

G10h4ck committed:
Improve API

Manually expose /rsFiles/getFileData to stream/preview files
Automatically expose a bunch of methods via JSON API
Implement serial_process for std::pair

G10h4ck committed:
Suggest shallow submodules for jsonapi dependencies

G10h4ck committed:
Avoid cloning huge unused restbed openssl dependency

G10h4ck committed:
Expose more RsFiles via JSON API

FileInfo remove dead static members

G10h4ck committed:
Enable JSON conversion for RsGxsImage via MemBlockProxy

hunbernd commented:
Additionally generate OpenAPI/Swagger definitions.
It makes possible to generate client side code for various languages.

G10h4ck committed:
Fix location creation via JSON API

To complete location creation login is needed

G10h4ck committed:
Merge branch 'csoler/v0.6-FT' into jsonapi

Needed to exopse JSON API for file search

sehraf committed:
convert values properly to string

sehraf committed:
make RsGroupInfo serializable

G10h4ck committed:
Fix warning #pragma once used in main file

Probably a copy paste error made during license cleanup

sehraf committed:
fix deserialisation and improve PRINT job

G10h4ck committed:
Expose /rsFiles/turtleSearchRequest via JSON API

Expose new async C++ API RsFiles::turtleSearchRequest with callback
Modernize TurtleFileInfo serialization code keeping retrocompatibility

G10h4ck committed:
Merge pull request #2 from sehraf/pr_improve_json

Multiple fixes on json handling by sehraf

G10h4ck committed:
Made JSON API port configurable from retroshare-gui

Use port used by webui + 2

G10h4ck committed:
Add JSON API support to retroshare-nogui

G10h4ck committed:
Add support for default parameters value in jsonapi-generator

G10h4ck committed:
Improve JSON API documentation

G10h4ck committed:
Improve documentation

csoler commented on jsonapi-generator/README.adoc:
```diff
@@ -0,0 +1,287 @@
+== How to use RetroShare JSON API
+
+Look for methods marked with +@jsonapi+ doxygen custom command into
++libretroshare/src/retroshare+. The method path is composed by service instance
```
Plz cite one particular method here by pasting its header from libretroshare. Maybe rsgxschannels.h?

G10h4ck replied:
Done ;)
Thanks!

csoler commented on jsonapi-generator/README.adoc:
```diff
@@ -0,0 +1,287 @@
+This may result a little tricky as language parsing or other adevanced things
+may be required.
+
+Hope this dive is useful for you +
```
you may want to remove these 3 lines ;-)

G10h4ck replied:
Well that is left there just as historical record on how this idea came out ;)

csoler commented on libretroshare/src/retroshare/rsnotify.h:
```diff
@@ -206,44 +206,42 @@ class RsNotify
 };
 
-class NotifyClient
+struct NotifyClient
```
Because this struct as some methods you might want to keep it as a class. As we discussed already, we should keep "struct" for pure data structures without functionality.

G10h4ck replied:
Done! Thanks!

csoler commented on libretroshare/src/rsserver/rsinit.cc:
```diff
@@ -1929,3 +1934,88 @@ int RsServer::StartupRetroShare()
+	rsNotify->setDisableAskPassword(false);
+	if(ret != RsInit::OK) return ret;
+	if(RsControl::instance()->StartupRetroShare() == 1) return RsInit::OK;
+	return RsInit::ERR_UNKOWN;
```
It seems to me we should clear the cached password here right after LockAndLoadCertificates(), since it was passed as a parameter.

G10h4ck replied:
Done! Thanks!

csoler commented on libretroshare/src/serialiser/rsserializer.h:
```diff
@@ -226,7 +220,7 @@ struct RsGenericSerializer : RsSerialType
 		 *  nested RsSerializable */
 		SerializeContext(
-		        uint8_t *data, uint32_t size,
+		        uint8_t* data = nullptr, uint32_t size = 0,
```
make sure that if only data is passed to SerializeContext, an error is printed out, otherwise you end up with some memory and no size for it.

G10h4ck replied:
Done! Thanks!

csoler commented on libretroshare/src/serialiser/rsserializer.h:
```diff
@@ -261,6 +255,12 @@ struct RsGenericSerializer : RsSerialType
+	 * to continue even if some field is missing (or incorrect), this way the
+	 * API is more user friendly as some methods need just part of the structs
+	 * they take as parameters. */
+	static const SerializationFlags SERIALIZATION_FLAG_YIELDING;       // 0x0008
```
This should obviously be restricted to JSon only. Using this in the binary serialization is very risky. Can you somewhere check this for consistency?

G10h4ck replied:
Done! Now serialization context check for usage of this flag on binary contexts and print an error and stacktrace if that happens

Thanks!

csoler commented on libretroshare/src/serialiser/rstypeserializer.cc:
```diff
@@ -328,6 +335,49 @@ bool RsTypeSerializer::from_JSON( const std::string& memberName,
 }
 
+template<> /*static*/
+uint32_t RsTypeSerializer::serial_size(const double&)
```
why is this needed? Either we try to serialize doubles, then we need this, or we don't, then we dont need this either. Why would JSon need doubles where the binary API doesn't? Also if necessary, serializing doubles is quite possible. I did it for float32 already I think.

G10h4ck replied:
We need this for JSON because some parameter type of some method exposed in the JSON API use double, binary serialization is not needed ATM because that kind of type is never exchanged between retroshare instances, I don't know the math behind floating point binary serialization so i haven't implemented it as is not needed ATM, but if you want to contribute that you are welcome to do that ;)

csoler commented on libretroshare/src/serialiser/rstypeserializer.h:
```diff
@@ -643,28 +732,32 @@ struct RsTypeSerializer
 
 			member.serial_process(j, lCtx);
-			ctx.mOk = ctx.mOk && lCtx.mOk;
+			ctx.mOk &= lCtx.mOk;
```
Strictly speaking this is dangerous, unless you make sure by design that ctx.mOk is always 0 or 1, which might depend on the definition of true and false and on the particular implementation and on the value you put into ctx.mOk (some function return integers).  !4 is false, !2 is false as well, but !(4 & 2) is true. Normally you'd need &&= but I'm not sure this exists.

G10h4ck replied:
Because ctx.mOk is a member of type bool in C++ (not in C) it is guaranted to be either true or false, so is is safe in that case, but i agree that the && variant is usually safer, and sadly &&= doesn't exists (we would benefit of that in other places especially because && is short circuit while & is not)

csoler commented on retroshare-gui/src/gui/gxschannels/CreateGxsChannelMsg.cpp:
```diff
@@ -449,11 +450,10 @@ void CreateGxsChannelMsg::addAttachment(const std::string &path)
 
 	FileInfo fInfo;
-	std::string filename;
+	std::string filename = RsDirUtil::getTopDir(path);
```
getTopDir() returns the file name??? Either a typo, or the name of the function is really not well chosen.

G10h4ck replied:
At first glance it seems to me that that function just split the path by / and return the last piece, so in case of a file path it is a file name, in case it is a dir path is the last dir name, but i agree that the function name is at least misleading...

G10h4ck committed:
Improve JSON API documentation following Cyril suggestion

G10h4ck committed:
Make NotifyClient class again

G10h4ck committed:
Friendly android-service build error if wrong compile options

G10h4ck committed:
RsLoginHelper::attemptLogin clear cached passphrase after usage

Thanks Cyril for suggestion

G10h4ck committed:
Add more consistency checks to serialization code

As suggested by Cyril
